### ISSUE TITLE
Weird (incorrect?) error handling behaviors

### ISSUE THREAD

**Issue by dieppe** [yelouafi/redux-saga#441]
Hello,

We've been enjoying using `redux-saga` for a month now, and we're quite happy with it (so, thank you for that 😄 ). 

There is something that we cannot figure out though: error bubbling and how to properly handle it in the root saga. (our goal is to have our root saga handle all the uncaught errors and restart the app when that's the case)

As a disclaimer: I don't know if all the following is expected behavior or not (I guess not). But even if it is, it is really confusing and we could not find anything in the docs related to that.

Anyway, here it goes:

I've noticed the following error handling (/bubbling) patterns:

```js
try {
  yield fork(...);
}
catch (error) { ... }
```

```js
try {
  const task = yield fork(...);
  yield task.done;
}
catch (error) { ... }
```

and 

```js
const task = yield fork(...);
try {
  yield task.done;
}
finally {
  const error = task.error();
}
```

I'll post different scenarii, all quite similar, but with different error bubbling behaviors (except scenario 2, for which I found none) .

(all these have been run against `redux-saga@0.11.0` -- transformed with `babel@6.10.1` and react native packager)

---

**Scenario 1**: Immediately throwing in a forked function
```js
// this is not a generator but fork expect a generator or a "normal" function that returns a Promise.
function immediatelyFailingFork () {
  throw new Error('immediatelyFailingFork');
}

export default function* root () {
  const error = yield fork(immediatelyFailingFork);
  console.log(error); // logs the error properly
}
```

Note: we don't fork a generator, but a function.
Expected behavior: a task is returned by `fork`.
Actual behavior: the error is returned by `fork`.

---

**Scenario 2**: Throwing in a function forked from a forked generator
```js
function immediatelyFailingFork () {
  throw new Error('immediatelyFailingFork');
}

function* launcher () {
  yield fork(immediatelyFailingFork);
}

export default function* root () {
  yield fork(launcher);
}
```

Note: same as Scenario 1, but we put the `fork` of the throwing function inside another `fork`.
Excepted behavior: Error bubbling to the `root` saga.
Actual behavior: The error is swallowed completely. No "uncaught at root" error is logged by `redux-saga` (Seems only logical since no task is returned in scenario 1, hence no error propagation can take place).

For this scenario, I tried the following error handling strategies:

**2.1**
```js
export default function* root () {
  try {
    yield fork(launcher);
  }
  catch (error) {
    console.log(error); // never called
  }  
}
```

**2.2**
```js
export default function* root () {
  const task = yield fork(launcher);
  try {
    yield task.done;
  }
  catch (error) {
    console.log(error); // never called
  }
  finally {
    console.log(task.error()); // logs undefined
  }
}
```

**2.3**
```js
export default function* root () {
  const task = yield fork(launcher);
  task.done.catch(error => console.error(error)); // never called
}
```

---

**Scenario 3** Immediately throwing in a forked generator (similar to 1)
```js
function* immediatelyFailingForkGenerator () {
  throw new Error('immediatelyFailingForkGenerator');
}

export default function* root () {
  try {
    yield fork(immediatelyFailingForkGenerator);
  }
  catch (error) {
    console.error(error); // logs the error properly
  }
}
```

Note: Similar to scenario 1, but instead of throwing in a `fork`ed function, we throw in a `fork`ed generator.
Expected behavior: (from some comments in the issues) use `task.done` to catch the error.
Actual behavior: no `Task` is returned from `fork`, an error is thrown instead.

---

**Scenario 4** Throw after a put in a forked generator
```js
function* failingAfterPutFork () {
  yield put({ type: 'AN_ACTION' });
  throw new Error('failingAfterPutFork');
}

export default function* root () {
  // we can put this into the try block, with no change, since the error is not thrown at this point
  const task = yield fork(failingAfterPutFork);
  try {
    yield task.done;
  }
  catch (error) {
    console.error(error); // never called
  }
  finally {
    console.error(task.error()); // logs the error properly
  }
}
```

Note: same as scenario 3, but this time we call `put` before throwing.
Excepted behavior: same as 3.
Actual behavior: a `Task` is returned by `fork`, but no error is thrown inside `task.done`. Instead we need to use `task.error()` in the `finally` block. Also, note that an "uncaught at root" error is logged by `redux-saga`, which should not be the case (but reflects the actual behavior).

---

**Scenario 5** Throw after a put in a forked generator in a forked generator
```js
function* failingAfterPutFork () {
  yield put({ type: 'AN_ACTION' });
  throw new Error('failingAfterPutFork');
}

function* launcher () {
  yield fork(failingAfterPutFork);
}

export default function* root () {
  try {
    yield fork(launcher);
  }
  catch (error) {
    console.error('in catch', error); // logs the error properly
  }
}
```

Note: same as scenario 4, but with an intermediate "launcher" that `fork` the throwing generator.
Behavior: same as scenario 3.

---

**Scenario 6** Throw in forked generator after a call in a forked generator
```js
function* immediatelyFailingForkGenerator () {
  throw new Error('immediatelyFailingForkGenerator');
}

function* launcher () {
  yield call(() => new Promise(resolve => {
    resolve();
  }));
  yield fork(immediatelyFailingForkGenerator);
}

export default function* root () {
  const task = yield fork(launcher);
  try {
    yield task.done;
  }
  catch (error) {
    console.error('in catch', error); // never called
  }
  finally {
    console.error('in finally', task.error()); // logs the error properly
  }
}
```

Note: this seems to be a variation on scenario **4** (edited, was 3, my mistake), as we see the same behavior.

---

It seems there are a three types of behavior (not counting scenarii 1 and 2) that can be encountered just by `fork`ing in some ways. Some might be bug, some might be related to 'synchronous' error throwing. I didn't have the time to look into the code more than a few minutes (and won't have the time to, sadly). 

Anyway, please let me know what I can do, even if as said, we're on a tight pre-release schedule, so time is scarce 😉

**Comment by yelouafi** [comment]
I'll look at this more closely, I think it's due to an incorrect handling in the lib of the sync forks. For example in Scenario 1, it's clearly a bug, since the fork shouldn't return the error but has to throw it (Actually [in this place of the code](https://github.com/yelouafi/redux-saga/blob/master/src/internal/proc.js#L441) we need to put `cb(err, true)` to indicate that this is an error).

But first le me indicate what the expected behavior should be in both sync and async cases (I'm planning to add this in the docs in a 'redux-saga fork model').

When using a `call` a generator should be able to catch errors from the called function/generator

```js
try {
  yield call(funcorGen, ...)
} catch(err) { ... }
```

Uncaught errors should bubble up *the call chain* (ie to `call`s from parents) just like in normal JavaScript code.

`fork` effects has different semantics. Observe this code for ex.

```js
function* parent() {
  try {
    yield fork(funcorGen, ...)
  } catch(err) { ... } // Invalid, you can't expect to catch errors from forked tasks
}
```

The reason you can't catch errors from forks is that the parent is not blocked and waiting for the fork but continues its execution as soon as the fork is performed; when a fork throws some error the parent Generator could have moved past the try catch block

```js
function* parent() {
  try {
    yield fork(funcOrGen, ...)
  } catch(err) { ... } 
  // when `funcOrGen` task throws, the Generator could be already here
  yield fork(anotherFuncOrGen) 
}
```

So the rule of thumb is **you can't catch errors from forks, but they still bubble up**. So how do Error propagation from forked tasks work?

>Note for all readers: If you dont want to dig deeper into this, you can just remember the rule above and skip all of this.

In the pre 0.10 versions all errors from forked tasks were swallowed. Just like in raw Promises and async functions now. 0.10 release introduced a new concept of *attached fork* [see release notes](https://github.com/yelouafi/redux-saga/releases/tag/v0.10.0) : a parent con fork one or multiple tasks, say for example we have a saga which at some time forked 2 tasks, so now we have 3 tasks evolving in parallel

- The 'main task' is the parent who forked the other tasks
- The 2 forked tasks

You can view those tasks as 3 parallel tasks belonging to the same execution context or *execution tree*. Each one of the 3 tasks can catch its own errors, but can' t catch errors from its siblings. When an uncaught error is raised inside one of the 3 tasks, the whole execution tree is aborted: the errored task is of course aborted, and the sibling tasks, if they are still running, are cancelled (including all their respective subtasks, effectively killing all the nested execution trees).  

After an execution tree is aborted, the error bubbles up to the parent of the main task. If this parent was calling the main task using `call` (or `join`) then it should be able to catch that error, but if it was using `fork` the same mechanism happens: the whole execution tree corresponding to that parent (its own task and all forked tasks) is cancelled and the error is raised again up to the root.

To summarize:

- A Generator, along with all its forked tasks, form an execution tree (forked tasks can also have their own nested execution trees). The generator forking the other tasks is called the *main task*, the main task can catch errors only from generators/functions invoked using a blocking call (`call` or `join`).

- The execution tree terminates when the main task's body terminates as well as all forked tasks (all nested execution trees should terminate)

- Cancelling the main task will cancel the whole execution tree

- Any uncaught error from a task (main tasks or one of the forks) in the execution tree causes the whole execution tree to abort. The error bubbles up to the parent of the main task (and the same rules applies to this parent).

So in relation to the above scenarios, it seems (I have still to look more closely) that this is caused by an inconsistent handing of the sybchronous forks (ie which terminates synchronously). [in the actual code](https://github.com/yelouafi/redux-saga/blob/master/src/internal/proc.js#L479) the sync errors from forks are reported to the main task, but they should behave like the async forks and cause the whole execution tree to abort. And since the fork aborts synchronously, we can't expect the fork call to return a task but immediately aborts the whole execution tree including the main task.

OTOH, *detached forks* , created using the `spawn` Effect behave like the pre 0.10 forks; errors raised from spawned tasks are lost unless you handle them explicitly (catching in the `done` promise of the task, of `join`ing a spawned task). This is pretty much the behavior of the raw promises and async functions: the behavior of error propagations depend on *whetger* and *when* you attached the error handler to a promise chain (before or after the error was raised).

The structured model of attached forks, may sound more restrictive and less flexible, but It's this structure and the rules above which makes things like Server Side Rendering possible with the same Saga code running both on the client and the server.

**Comment by dieppe** [comment]
Thanks for this very complete answer.

I have some questions though. I do understand that when we call fork, the parent task continues its execution. That's why in our case, we have one indirection: root (forks) launcher (forks) all-long-running-sagas, so we can do the following and (that's where I'd like to know if we assume right) catch all errors:

```js
function* root () {
  const task = yield fork(launcher);
  try {
    yield task.done; // this will throw if an error bubbles (?)
  }
  catch (error) { }
}
```

Another question is, right now, the fork must be itself in the `try..catch` as presented in scenarii 3 and 5, but it should not be right (that's what you referred by sync forks?)?

Anyway, for now, that's what we have in place, and it seems to catch all errors, even if we sometimes see the "uncaught at root" log (I'm putting that here as it might interest someone):

```js
function* root () {
  let launchTask;
  let errorAlreadyHandled = false;
  try {
    launchTask = yield fork(launchSagas); // launch all long-running sagas
    yield launchTask.done; // wait for the task, so that if the done promise rejects, we can catch the error 
  }
  catch (error) {
    errorAlreadyHandled = true;
    yield call(handleUncaughtError, error);
  }
  finally {
    if (!errorAlreadyHandled && launchTask.error()) {
      yield call(handleUncaughtError, launchTask.error());
    }
  }
}
```

**Comment by dieppe** [comment]
OK, disregard my last comment (seems I didn't understand your answer fully 😅  ).

I'll probably repeat what you said in your comment, but I want to be sure I'm not mistaken:
When an error is thrown (and not caught) inside a task T, the whole execution subtree starting at T is aborted. If T is itself part of an execution tree, the error is bubbled up to the parent task P of T. Now, I'd like to know for sure: can P handle the error and stop the cancellation from bubbling?

**Comment by yelouafi** [comment]
>can P handle the error and stop the cancellation from bubbling?

Yes but only if P is using a blocking call (`call`, `join` or a composite form of them like `[]` or `race`). If P is invoking T using `fork` the same mechanism happens: P is aborted.

to get a better intuition, you can think of a saga + its forks as a *dynamic parallel* effect. In a *static parallel* effect

```js
function* saga() {
  try {
     yield [call(s1), call(s2)]
   } catch (err) { ... }
}
```

the parallel effect ` [call(s1), call(s2)]` will terminate only after all effects inside terminate. And will abort as soon as one effect inside the array raises an uncaught error. And note `s1` for example can not catch errors from its sibling `s2`. But `saga` can catch errors from the array because `yield [call(s1), call(s2)]` is a blocking call.

Now with forks

```js
function* childSaga() {
  while(true) {
    yield take(ACTION)
    yield fork(s)
   } 
}

function* saga() {
  try {
    yield call(childSaga)
  } catch(err) { ... } // ok, can catch errors b/c it's using a blocking call
}
```

You can think of `call(childSaga)` as `[call(s1), call(s2)]`. But now instead of a fixed list, we have a dynamic of tasks (because we dont know in advance how many tasks w're going to call, it depends on the number of dispatched actions):

```js
yield [main, call(s1), call(s2), ...]
```
The 1st one is the *main task* `childSaga` itself which runs the `while(true)` loop and forks other tasks dynamically, and the rest are the tasks forked by main. The list may grow dynamically as new tasks are forked (or shrink as existent tasks are terminated). And same as the static parallel effect, any error from one of those tasks will cause the whole 'dynamic array' to abort.

**Comment by dieppe** [comment]
All right, I get what you mean. I'll try to be more specific in my question, because I feel there is some part of it still unanswered (in light of your last comment though, I think it's more theoretical than practical, since I now see how to handle errors properly in our codebase).

The doc says that a fork effect does return a task to the calling generator (which is not always true in case of synchronous error, but that seems to be a bug):

```js
function* main () {
  const task = yield fork(...);
}
```

My understanding of it was that this task object represents the fork, and can be used to "wait" on the fork by doing the following:

```js
function* main () {
  const task = yield fork(...);
  yield task.done;
}
```

My impression was that with this bit of code one could "emulate" a `call` effect while using a fork. I now see that it's useless since using a `call` does the exact thing I was looking for, but this is what has been bugging me and why I started this issue, because right now this code does not work the same depending on what happens in the fork.

To summarize, I do understand forked tasks run in parallel and there is no way to "wait" for them implicitly, but I thought redux-saga provided a mechanism to explicitly wait on the `task` object representing the `fork` execution, and also that this provided us with the opportunity for the parent task to handle the error of its forked task before the parent itself is canceled.

Hence, I would expect the following to be true:
```js
function* main () {
  yield fork(t1);
  yield fork(t2);
  // if t1 or t2 throws, the whole execution tree should be aborted, and redux-saga should warn about uncaught error
}

function* main () {
  yield fork(t1);
  const task = fork(t2);
  try {
    yield task.done;
  }
  catch (error) {
    // if t2 throws, then t1 still runs and main can recover
    // if t1 throws, then the whole tree is aborted and redux-saga warns
  }
}
```

**Comment by yelouafi** [comment]
I'm not sure to understand

- If you want to wait for a task immediately after forking it you can simply use `call`. 

- If you want to wait for a specific task `t` you can use `yield join(t)`

- If you want to handle errors from all forks in a saga and prevent that Saga from being aborted (I suspect this is what you want). You can either handle the errors inside the forked tasks themselves, or if you want to keep it a separate concern, you can wrap all forks in a generic handler

```js
function* wrapFork(saga, args = []) {
  try {
    return yield call(saga)
  } catch(err) { 
     ...
   }
}

function* parent() {
  while(true) {
    yield take(action)
    yield fork(wrapFork, saga)
  }
}
```

But if you wanted my 'very opinionated way' for handling errors, you shouldn't rely on try/catch to handle errors in all JavaScript and use explicit return values for 'expected (business)' errors. Then let the 'unexpected errors' (bugs) bubble up and abort the program (catch them at the top saga to send a crash report)

**Comment by dieppe** [comment]
You are totally right in the fact that we should use `call` (in fact I already made the change). And we do handle the errors as you proposed at the end of your comment. 

In fact, this all started because we wanted to implement that and catch unhandled errors (crashes) in the top saga. I read about using `task.done.catch` for forked task [here](https://github.com/yelouafi/redux-saga/issues/250#issuecomment-207856839).
I did choose the wrong way of implementing the 'catch-all-crashes-top-saga' using `fork` and the`task.done.catch` technique instead of `call`. It did not work, hence the issue being opened. 

Now, all that is very clear to me, but since I'm curious (and like when things are consistent) and want to know a bit more how `redux-saga` works, I wanted to know if it *could have been* done  (not if it *should* be done) using

```js
const task = yield fork(launcher);
try { yield task.done; }
catch { // report crash }
```

and it does not work simply due to a bug, or if I'm mistaken and that could have never worked anyway.

Sorry if I was not clear before (text-only communication is hard): 
- I'm not interested in a solution anymore (since you provided that already -- and by the way, thanks a lot for that 😉 ), 
- I did understand the solution and the execution model of `fork`,
- but I just wanted to know if *in principle* the above code *should* have worked.

**Comment by yelouafi** [comment]
sorry for the delay. The above should not work. since you can't catch errors from forked tasks.

BTW, instead of `yield task.done`, it's better to use `yield join(task)`.

FYI, I've already implemented the fixes and also added initial docs for the fork behavior. To be uploaded shortly with a new release
